core/src/main/java/com/scalar/db/transaction/consensuscommit/ParallelExecutor.java

```diff
@@ -45,6 +52,49 @@ public void prepare(List<ThrowableRunnable<ExecutionException>> tasks) throws Ex
     executeTasks(tasks, config.isParallelPreparationEnabled(), false);
   }
 
+  public void validate(List<ValidationTask> tasks)
+      throws ExecutionException, CommitConflictException {
+    List<Future<?>> futures;
+    if (config.isParallelValidationEnabled()) {
+      assert parallelExecutorService != null;
+      futures =
+          tasks.stream()
+              .map(
+                  t ->
+                      parallelExecutorService.submit(
+                          () -> {
+                            t.run();
+                            return null;
+                          }))
+              .collect(Collectors.toList());
+    } else {
+      futures = Collections.emptyList();
+      for (ValidationTask task : tasks) {
+        task.run();
+      }
+    }
+
+    for (Future<?> future : futures) {
+      try {
+        Uninterruptibles.getUninterruptibly(future);
+      } catch (java.util.concurrent.ExecutionException e) {
+        if (e.getCause() instanceof ExecutionException) {
+          throw (ExecutionException) e.getCause();
+        }
+        if (e.getCause() instanceof CommitConflictException) {
```

Comment:
> The reason for not using executeTasks is to handle CommitConflictException?

Yes, correct.

> I'm sure you've thought about it but you don't like using executeTask for this by adding the CommitConflictException handling?

If we add the CommitConflictException handling to `executeTasks`, the other methods (prepare()/commit()/rollback()) need to handle `CommitConflictException` as well, they can't throw `CommitConflictException` though. But, it will reduce the code redundancy. I don't know which is better, honestly. Okay, let me add a commit to add the CommitConflictException handling to `executeTasks`, and let's compare and check which is better. Thanks.

Reply:
@feeblefakie  I added a commit to add the CommitConflictException handling to the `executeTasks` method. After this change, we need to ignore `CommitConflictException` in prepare()/commit()/rollback(). Please take a look!